### DISCUSSION TITLE
Mouse selection style differs slightly from NSTextView

### DISCUSSION THREAD

**Discussion by humblehacker** [krzyzanowskim/STTextView#46] [Questions]
I've discovered a small difference in the selection styles between `NSTextView` and `NSTextField`. Specifically, when Shift-clicking before some already selected text, `NSTextView` will extend the selection, while `NSTextField` will replace the selection.

To illustrate:

_brackets `[` and `]` mark the selection, while `|` marks the shift-click point._

|    |    |
| --- | --- | 
| before | `12\|345[678]9` |
| `NSTextView` | `12[345678]9` |
|`NSTextField` | `12[345]6789` |

As it turn out, `STTextView` follows the `NSTextField` behavior of replacing the selection rather than the `NSTextView` behavior of extending the selection. So perhaps if `STTextView` is meant as a replacement for `NSTextView` its selection style should match? Or maybe made configurable?

**Comment by krzyzanowskim** [comment]
Thank you for bringing that detail to my attention! That's an interesting detail. You're right that STTextView does not follow the NSTextView selection behavior.

There's more to that story. The selection seems to vary between editors.

- Safari (NSTextView style)

  https://github.com/krzyzanowskim/STTextView/assets/758033/36dcc4ce-ff33-4876-b335-4e95a7ad2a02

- Firefox (NSTextField style)

    https://github.com/krzyzanowskim/STTextView/assets/758033/4a40d3f7-eac9-4941-af9c-865ecb7f6531

- VSCode (NSTextField style)

  https://github.com/krzyzanowskim/STTextView/assets/758033/30faafbd-e85f-4aa1-8e11-23ec3cb6afc4

- BBEdit (modified NSTextView style)

  https://github.com/krzyzanowskim/STTextView/assets/758033/9938a0ad-5b3f-4c98-a26a-eea27d5e50e6

**Comment by humblehacker** [comment]
While diving deeper into this, I've come to believe that the `NSTextField` style is really just a bug, and that the `NSTextView` style is more correct. The difference essentially comes down to where the new selection is anchored. For `NSTextField`, the anchor is always the beginning of the selection, while for `NSTextView` the anchor is the midpoint of the selection. The latter makes more sense to me. It's more predictable, and works the same regardless of which side of the selection you click on. My vote would be to have `STTextView` mirror the behavior of `NSTextView`.

**Reply by humblehacker** [reply]
I could be wrong on the "bug" aspect. Just for fun I tried some text selection in gnome on Linux and in Windows. The most common selection style appears to be the `NSTextField` style, though I did see the `NSTextView` style on Firefox on Linux only.

I still think the `NSTextView` style is preferable, unless someone can explain how the `NSTextField` style is superior.

**Reply by krzyzanowskim** [reply]
> while for NSTextView the anchor is the midpoint of the selection

what do you mean by midpoint here? we may be thinking about different things. for me anchoring in the midpoint does not result in extending either way

**Reply by humblehacker** [reply]
Sorry, that wasn't entirely clear - maybe _anchor_ is the wrong term. It's just that the midpoint is used to determine which way to expand/shrink the selection. Click anywhere to the left of the midpoint, the selection will expand/shrink to the new point on the left, depending on whether you click inside or outside the selection. And vice-versa for clicking to the right of the midpoint.

**Comment by krzyzanowskim** [comment]
> My vote would be to have STTextView mirror the behavior of NSTextView

I agree on that. I'm surprised TextKit doesn't do it automatically for me. Or I'm using it wrong. either way, something to address

**Comment by krzyzanowskim** [comment]
Adding **TextEdit** app to the mix, that is following the _NSTextField style_

https://github.com/krzyzanowskim/STTextView/assets/758033/5e11f7b2-c1d2-4a7a-8050-aff9be11e709

That one makes me confused even more.

![Screenshot 2024-02-10 at 02 20 59](https://github.com/krzyzanowskim/STTextView/assets/758033/c471a184-f569-420b-8e84-5a15fe8ec3a9)

TextEdit app uses **NSTextView**; however, the selection behaves like _NSTextField style_. TextEdit app also uses NSTextView in TextKit2 mode. I start to think that is the new default and it behave different depends whether NSTextView uses TextKit1 or TextKit2 setup [I didn't test that theory]

**Comment by tempelmann** [comment]
TL;DR: While initial click-drag selection should always work around the anchor, even if moving the mouse first to the right and then to the left (while still dragging), additional shift-clicks should probably expand on either side, without clearing previous selection (though it would be nice if that where to be done only with an additional modifier). Compare with cursor-selection. Special case: Starting with word-selection (from double click). Also: **Let's consult Inside Macintosh**.

I like to point out that it's usually important HOW you selected the text initially. If you mouse-click-and-drag, then there's a start point, and the classic (original) rule was that if you extend the selection, you always replace the selection with the points between your original click and the new shift-click. Compare this also to selecting with cursor keys and shift-cursor: Move the cursor to a point in a word, the shift+forward, then shift+backward. You'll see that it keeps anchoring from the point where you started. You'd expect the same behavior from mouse shift-selection.

A special case is selecting a word by double clicking. Still, the same behavior should be kept, only that the start point is the entire word, i.e. the word always remains selected entirely, but shift-clicking on either far side would then select anything between the shift-click and the entire word.

Still, all this means: Shift-clicking left side, then right side, should never extend the selection on BOTH sides. That's just not conforming with the behavior when selecting with keys, and makes it hard to predict. Of course, sometimes a user WANTS to extend the selection on both ends. For that, another method is needed, though (e.g. by pressing another modifier).

OTOH, one could also argue this: Expanding a selection with shift-cursor_keys is behaving the same as click-shift-drag selection: If you click, then drag to the right, it expands only to the right and when you keep dragging, going to the left of the initial click, the selection expands only to the left. Contrary to that, once this drag-selection is finished, _another_ shift-click could then always expand the selection, to either side, without clearing the initial selection. The difference to the initial drag-selection is that you are already finished with that (this doesn't work with keyboard selection though, because you cannot move the cursor to another place without clearing the selection automatically). In fact, that feels more natural to me, indeed.

Also consider how you expand selections in a list (TableView): If you select, then shift-click above or below, it also replaces the entire selection with the rows between your initial row and the shift-clicked row. I.e. no expansion. You can only expand by using the cmd modifier (but, at least in Finder, you cannot shift-cmd-drag, but only shift-cmd-click for single row selection expansion).

So, the consensus seems to be that: Shift-click always re-selects between the initial point and the shift-click. If you can indeed expand the previous selection, that may be a poor attempt by some programmers to allow multi-expansion without basing that on a common specification on how to do that consistently. I mean: **There's nothing about multi-expansion behavior in Inside Mac, IIRC**. It's a poor interpretation that's been causing this mess you're pointing out now. If those who started this expansion method had required an additional key, e.g. cmd, to be held alongside, this could have been avoided. Boo to the "smart" devs that did that without consulting the UX team that Steve fired. ;)

**Reply by krzyzanowskim** [reply]
You're saying Safari has it wrong?

**Reply by tempelmann** [reply]
Well, I think that it's practical to extend the selection with shift-clicks at both ends, but it's not what Raskin worked out originally, I think. There were other rules back then, and ideally the new rules, which are more practical, should be announced in a wider fashion so that everyone adopts them instead of following the original rules (which some still do, and I would as well hadn't I given this some more thought). And the new rule should also applied to list view and grid view selections (in the latter case, Cocoa's `NSCollectionView` doesn't do any selection handling, meaning I had to code that all myself, which took many hours to get it "right", hence I've given this some thought in the past. Correct me if I'm wrong, of course :)

**Comment by krzyzanowskim** [comment]
On top of what was said above, I tried the `NSTextView` setup in macOS 14 (Sonoma). Turned out the selection behavior differs depending on what version of TextKit it uses.

**TextKit 1** is the selection extending like @humblehacker reported here
**TextKit 2** is the selection extending like in STTextView (replacing selection) 

Hard to tell whether that is intentional or not, but that at least sort out the mystery for me.

https://github.com/krzyzanowskim/STTextView/assets/758033/44f6a816-aee8-4b77-9ff0-1df553f057ae

**Comment by krzyzanowskim** [comment]
To summarize my thoughts. I believe the current behavior is correct/expected in the long run. It is a new behavior derived from the TextKit 2, and it spread across the macOS, too. NSTextView, when using TextKit 2 setup, also follows that new way of selection. If anyone wants to provide implementation to support fallback behavior, I'm ok with that. However, I don't plan to invest my time in it in the foreseeable future by myself.
